### PR TITLE
feat: Add support for OR in the pipeline

### PR DESCRIPTION
**Pull request by arv**
This does not yet build the AST.

It adds three new operator

## BranchOperator

The branch operator fans out.

```
     A
   / | \
  B  C  D
```

All messages received at A are forwarded to `B`, `C` and `D`.

## ConcatOperator

Is the reverse of the `BranchOperator`. It has multiple inputs and a single output.

```
  B  C  D
   \ | /
     A
```

All messages received are forwarded to A

## DistinctOperator

This operator "removes" "identical" messages received in the same "transaction"

 - removes: It sums up the multiplicity. Then it collapses the multiplicity to either 1 or -1.
 - identical: Right now it compares the value by `===`. This will not work  since we need to do JSON equality to deal with different objects representing the same logical value.
 - transaction: The call to `run`

@tantaman There are some TODOs in the code that I would like some clarity in.

### CODE REVIEW

**Comment by arv** [comment]
TODO: Change `OperatorBase` to be `DifferenceStreamReader[]` and `DifferenceStreamWriter[]` instead of `DifferenceStreamReader[]` `DifferenceStreamWriter`

**Comment by tantaman** [comment]
Do we need a branch operator?

You can model this:

```ts
source.where(or(x=1, y=2))
```

like this:

```ts
const stream1 = source.filter(x=1);
const stream2 = source.filter(y=2);
const final = stream1.concat(stream2).distinct();
```

Without introducing `branch`

**Comment by tantaman** [comment]
> identical: Right now it compares the value by ===. This will not work since we need to do JSON equality to deal with different objects representing the same logical value.

The values flowing through the pipeline are rows and all rows _should_ be entities with an id. We can use the id for equality. Assuming this invariant still holds when we move this code out of rails.

I've implemented `join` and `group-by` to create a synthetic ID to preserve this invariant.
- the id of a group is the value of the set of fields grouped on
- the id of a joined row is id of all the other rows participating in the join

**Review by tantaman — CHANGES_REQUESTED** [review]

**Comment by arv** [comment]
> The values flowing through the pipeline are rows and all rows should be entities with an id. We can use the id for equality. Assuming this invariant still holds when we move this code out of rails.

We use bare `T` as a type param in **a lot** of places. We will need to change that to `T extends Entity` then. I'm fine with that but just wanted to make sure I understand before doing that.

File: src/zql/ast-to-ivm/pipeline-builder.test.ts [hunk 1]
```diff
@@ -95,5 +97,162 @@ test('Where', () => {
...
+        orderBy: [['id'], 'asc'],
+      };
+
+      const pipeline = buildPipeline(() => s.stream, ast);
+
+      // console.log(pipeline.toString());
+
+      const log: unknown[] = [];
+
+      // why in the world is effect not notified but debug is?!?!?!?!?!
```
Comment (tantaman) [inline src/zql/ast-to-ivm/pipeline-builder.test.ts:236]: so the problem is that `effect` only notifies on commit (intentional) but it doesn't retain values it has seen in the current transaction.

`or` -> `concat` -> `distinct` emits twice:
- Once for the left branch feeding into it
- Once for the right branch

If the row passes both `OR` conditions then:
1. The first time it emits a row,
2. The second time it does not emit a row

Effect throws out its old value and replaces it with nothing at step (2), causing effect to notify with empty results.

tldr: effect should retain all values it sees then notify with them on commit.

**Comment by arv** [comment]
Superseded by #56
